Making easy to use some features

just-do-halee opened the pull request.

just-do-halee committed:
easy to use CollisionPair on iter

just-do-halee committed:
easy to use KeyboardState chaining function caller

just-do-halee committed:
easy to use MouseState chaining function caller

just-do-halee committed:
[fixing] KeyboardState, MouseState: .._do method can be nested

CleanCut commented on examples/scenarios/car_shoot.rs:
```diff
@@ -132,7 +132,8 @@ fn game_logic(engine: &mut Engine, game_state: &mut GameState) {
         }
-        for label in [event.pair.0, event.pair.1] {
+
+        for label in event.pair {
```
❤️ Love it!

just-do-halee replied:
Hehe

Commit:
revert most of the examples

CleanCut reviewed (approved):
Cool! I think these are great as additional features!

I do not feel that these are _easier to use_ than the existing features, so I have reverted most of the examples, leaving a commented-out example in `examples/keyboard_state.rs` as a teaser. I feel like this is a nice "discoverable" feature for more advanced/ambitious students.

Question: I noticed that in your own conversion of the examples, you never needed to make use of the `&MouseState` or `&KeyboardState` that you passed to the closure. Are you certain that you don't want to remove that parameter? I wouldn't need it for the simple scenarios I lead students through, but perhaps you find it useful? If you would like to make that change, I'll wait for you. If you like it how it is, I'll go ahead and merge this now. Let me know!

Commit:
clippy leveled up

Commit:
update changelog

just-do-halee commented:
> Cool! I think these are great as additional features!

- Thank you So much! 😍 

> I do not feel that these are easier to use than the existing features, so I have reverted most of the examples, leaving a commented-out example in examples/keyboard_state.rs as a teaser. I feel like this is a nice "discoverable" feature for more advanced/ambitious students.

- I agree that `a nice "discoverable" feature for more advanced/ambitious students` Yeah, you are always a genius teacher. Thank you so much! And Thank you for commenting on it. It's really awesome! 😆

> Question: I noticed that in your own conversion of the examples, you never needed to make use of the &MouseState or &KeyboardState that you passed to the closure. Are you certain that you don't want to remove that parameter? I wouldn't need it for the simple scenarios I lead students through, but perhaps you find it useful? If you would like to make that change, I'll wait for you. If you like it how it is, I'll go ahead and merge this now. Let me know!

- ~~You're right! So I reverted the last commit once again but... then i realized that i commited it because of the timestamp `1:30` of `34. Tutorial: Mouse Input` in the `Ultimate Rust Crash Course 2`. ***`mouse_state`*** has ***location()*** and ***motion()*** and ***wheel()***. And i noticed that, How can i make nested conditions if i want to? Then also i was feel that using the borrowed state as an argument is much more idiomatic way to have functionality on Rust. Well, but also as a student, you are right! So I'm not sure what can i do. * The hidden problem is that we can not use `&mut Engine` inside that closure... But It's not bad noticing pros and cons on Rust functional programming, I was thinking.~~

---

Anyway, I love your all lectures on Udemy. Teacher Nathan!
I'd started learning Rust-lang 1 year and half months ago and it was from your first `Ultimate Rust Crash Course`.
I've been shocked how the teacher can explain and show the illustrations so, so, so easy to understand and then I became a fast rustacean learner thanks to you! Teacher Nathan.
I didn't `Ultimate Rust Crash Course 2` in that time, because there wasn't,
but `Hands-On Systems programming with Rust`.

I recently started `Ultimate Rust Crash Course 2`. 95% completed. And I can say that it is incredibly amazing lecture too!
Thank you for taking a closer look at my Pull-Request and commenting!

just-do-halee commented:
The &mut Engine is now available in the closures... And I removed the method name `_do` on the chainer(*_if().*).

The argument in the closure, `|&KeyboardState|` and `|&MouseState|`, I'll leave the choice to you.... My opinion is

> 1. It enables nested conditions.
> 2. You may need to use several `MouseState dependent methods` in the MouseStateChainer closure. (It can also occur later in the KeyboardState, we don't know)

just-do-halee committed:
[changed] _state.[just_]pressed[_any]_do(..) -> _state.if_().[just_]pressed[_any](..) now you can use &mut Engine inside the closure

Commit:
*Chainer -> *Chain, if_() -> chain()

CleanCut commented:
Alright, I left the reference parameters in.

I tweaked the names a little bit. I renamed `if_` to `chain()`, and I renamed `*Chainer` to `*Chain`. I think that's a bit more consistent. I also updated the changelog entry.

CleanCut commented:
Included in version `5.2.0`, just released. 😄
